File: src/bmsx/machine/runtime/frame/step.ts
import type { Runtime } from '../runtime';

export type RuntimeFrameStepResult = {
	previousTickSequence: number;
	tickSequence: number;
	tickAdvanced: boolean;
};

export function createRuntimeFrameStepResult(): RuntimeFrameStepResult {
	return {
		previousTickSequence: 0,
		tickSequence: 0,
		tickAdvanced: false,
	};
}

export function runRuntimeFrameStepInto(out: RuntimeFrameStepResult, runtime: Runtime, hostDeltaMs: number): void {
	const previousTickSequence = runtime.frameScheduler.lastTickSequence;
	runtime.frameScheduler.run(hostDeltaMs);
	const tickSequence = runtime.frameScheduler.lastTickSequence;
	out.previousTickSequence = previousTickSequence;
	out.tickSequence = tickSequence;
	out.tickAdvanced = tickSequence !== previousTickSequence;
}

File: machine/cpp/machine/runtime/frame/step.cpp
#include "machine/runtime/frame/step.h"

#include "machine/runtime/runtime.h"

namespace bmsx {

void runRuntimeFrameStepInto(RuntimeFrameStepResult& out, Runtime& runtime, f64 hostDeltaMs) {
	out.previousTickSequence = runtime.frameScheduler.lastTickSequence;
	runtime.frameScheduler.run(runtime, hostDeltaMs);
	out.tickSequence = runtime.frameScheduler.lastTickSequence;
	out.tickAdvanced = out.tickSequence != out.previousTickSequence;
}

} // namespace bmsx
